framework.cpp

```diff
@@ -0,0 +1,403 @@
+     set_t_l
+    >::type
+    sort_by_distance(set_t_l in, typename Metric::Rec_T center) { // only STP version now
+
+        set_t_l out (in.size());
+
+        auto cmp = [](std::pair<size_t, double> const & a, std::pair<size_t, double> const & b)
+        {
+            return a.second != b.second ? a.second < b.second : a.first < b.first;
+        };
+
+        std::vector<std::pair<size_t, double>> distances;
+        for (size_t i = 0; i<in.size(); i++) {
+            distances.push_back(std::make_pair(i, metric(in[i], center)));
+        }
+        sort(distances.begin(), distances.end(), cmp);
+        for (size_t i = 0; i<distances.size(); i++) {
+            out[i] = in[distances[i].first];
+        }
+        return out;
+    }
+    // */
+
+
+protected:
+
+    Metric metric;
+};
+
+
+
+
+// --------
+// derived classes
+
+
+template <typename rec_t>
+class EuclideanNoIter : public MetricBase<rec_t> {
+
+public:
+
+    double operator()(const rec_t & a, const rec_t & b) const  { // siuts both Blaze and STL
+        double sum = 0;
+        if (a.size() != b.size()) {
+            double dnan = std::nan("Input container sizes do not match");
+            return dnan;
+        }
+        for (size_t i = 0; i < a.size(); ++i) {
+            sum += (a[i] - b[i]) * (a[i] - b[i]);
+        }
+        return std::sqrt(sum);
+    }
+
+};
+
+
+
+template <typename rec_t>
+class EuclideanTypeSpecific : public MetricBase<rec_t> {
+
+public:
+
+    template <typename rec_t_l>
+    typename std::enable_if <
+     std::is_same<rec_t_l, rec_t>::value && determine_container_type<rec_t_l>::code == 1, // STL, we can use iterators
+     double
+    >::type
+    operator()(const rec_t_l & a, const rec_t_l & b) const { // STL overload, TODO add Blaze overload
+        if (a.size() != b.size()) {
+            double dnan = std::nan("Input container sizes do not match");
+            return dnan;
+        }
+        double sum = 0;
+        for (auto it1 = a.begin(), it2 = b.begin(); it1 != a.end() && it2 != b.end(); ++it1, ++it2) {
+            sum += (*it1 - *it2) * (*it1 - *it2);
+        }
+        return std::sqrt(sum);
+    }
+
+    template <typename rec_t_l>
+    typename std::enable_if <
+     std::is_same<rec_t_l, rec_t>::value && determine_container_type<rec_t_l>::code == 2, // Blaze, we can use Blaze matrx operations
+     double
+    >::type
+    operator()(const rec_t_l & a, const rec_t_l & b) const { // STL overload, TODO add Blaze overload
+        if (a.size() != b.size()) {
+            double dnan = std::nan("Input container sizes do not match");
+            return dnan;
+        }
+        auto diff = a - b;
+        double sum = blaze::sum(diff*diff);
+        return std::sqrt(sum);
+    }
+
+    template <typename rec_t_l>
+    typename std::enable_if <
+     std::is_same<rec_t_l, rec_t>::value && determine_container_type<rec_t_l>::code == 3, // Eigen, [] to access elements (or we can use Eigen-specific matrix operations)
```

Comment:
please help me to understand what is the problem with Eigen, as I see the latest version supports STL-compatible iterators

Reply:
there are iterators for Eigen objects inside STL containers, but Eigen object itself has no .begin() method and it's element can't be accessed in the way 
```cpp
for (auto it1 = a.begin(), it2 = b.begin(); it1 != a.end() && it2 != b.end(); ++it1, ++it2) {
            sum += (*it1 - *it2) * (*it1 - *it2);
        }
```

Reply:
in the new code i've posted to Mattermost i detect the Eigen type via additional specialization of my metafunction and access elements using [] operator